content/blog/2025-03-10-fixing-memory-leak/index.md

```diff
@@ -0,0 +1,76 @@
+the JVM was not terminated after an ETL but listened further for incoming ETL
+triggering. In principle the JVM should have only been restarted when new code
+was to be deployed, but we soon discovered that restarting [our Play app](https://github.com/hbz/lobid-resources/tree/master/web)
+just before the weekly fresh and complete ETL of > 24 million documents improved
+the performance and averted some hanging processes or crashes.
+
+We also have a monitor tool installed at our servers who checks if processes
+are terminated and restart those automatically. This happened often after some
+more ETL processes were invoked (the "daily updates").
+It was unclear why these crashes appear but by assigning more RAM and 
+(automatically) restarting the whole JVM after crashes the ETL process went
+stable enough.
+
+Since 2025 we started the [Rheinland-Pfälzische Bibliografie (RPB)](https://github.com/hbz/rpb/).
+Here, ETL processes are triggered _every time_ a record is changed or created.
+This resulted very fast in an
+[OutOfMemoryError](https://github.com/hbz/lobid-resources/issues/2121#issuecomment-2631355294).
+
+Using Intellij IDEA we can easily produce a picture which shows the ever
+increasing demand of memory:
+![Screenshot of the graph showing the increasing demand of memory](./intellijIdeaGraphShowingMemoryLeak.png)
+At the beginning (at the left) we see that the CPU (green colored) is heavily used
+for all classes are compiled and loaded. Then we see how the Heap Memory (blue
+colored) is filled and after some time there are some spikes, indicating that
+the Garbage Collector was able to free some unused memory.
+We can see how the memory
+consumption is ascending nonetheless. Then a long line appears, lacking all
+spikes. This happens when there is less and less memory which the
+garbage collector could release. We can see 5 peaks where the CPU heavily
+works - this is the garbage collector trying to find some piece of memory
+to free - without much avail. The app crashes shortly after that (not to be
+    seen).
+
+After [fixing the memory leak](https://github.com/metafacture/metafacture-core/commit/b32609307f75187a6a3822b8a951429c7fc924f3)
+the consumption of memory is normal, i.e. not ascending:
+![Screenshot of the graph showing normal consumption of memory](./intellijIdeaGraphShowingMemoryWithoutLeak.png)
+Every spike indicates that memory resources were freed, resulting in a stable
+rise and fall of CPU and memory usage. The memory leak is really fixed. 😀
+
+Fixing the memory leak in `metafacture` resolved some issues we've experienced
+at:
+- `lobid-resources`: daily updates sometimes aborted - although this was not so a big thing because our monitor scripts could "heal" the update process automatically (by restarting the app). However, the updates now don't take e.g. 4h (counting from triggering the update until the successful ETL) but 0.45m, which is way faster.
```

Comment:
```suggestion
- `lobid-resources`: daily updates sometimes aborted - although this was not such a big thing because our monitor scripts could "heal" the update process automatically (by restarting the app). However, the updates now don't take e.g. 4h (counting from triggering the update until the successful ETL) but 0.45m, which is way faster.
```